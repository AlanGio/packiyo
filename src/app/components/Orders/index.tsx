import React from "react";
import { Typography, Paper, Grid, Button } from "@mui/material";
import { OrderData } from "./orders.props";
import Link from "next/link";

const Orders = ({ orders }: { orders: OrderData[] }) => {
  return (
    <Grid container spacing={3} sx={{ mb: 20 }}>
      {orders.map((order: OrderData) => (
        <Grid item xs={12} sm={6} key={order.id}>
          <Paper elevation={3} sx={{ padding: 2 }}>
            <Typography variant="h6">{order.attributes.number}</Typography>
            <Typography>Status: {order.attributes.status_text}</Typography>
            <Typography>
              Shipping: ${order.attributes.shipping.toFixed(2)}
            </Typography>
            <Typography>Tax: ${order.attributes.tax.toFixed(2)}</Typography>
            <Typography>
              Discount: ${order.attributes.discount.toFixed(2)}
            </Typography>
            <Typography>Total: ${order.attributes.total.toFixed(2)}</Typography>
            <Link href={`/order/${order.id}`}>
              <Button variant="contained" color="primary" sx={{ mt: 2 }}>
                View Order
              </Button>
            </Link>
          </Paper>
        </Grid>
      ))}
    </Grid>
  );
};

export default Orders;
