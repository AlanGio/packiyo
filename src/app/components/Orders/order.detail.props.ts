export interface OrderDetailProps {
  type: string;
  id: string;
  attributes: Attributes;
  relationships: {
    customer: Relationship;
    shipping_method: Relationship;
    shipping_contact_information: Relationship;
    billing_contact_information: Relationship;
    order_channel: Relationship;
    shipping_box: Relationship;
    order_items: {
      links: {
        related: string;
        self: string;
      };
      data: RelationshipData[];
    };
    shipments: {
      links: {
        related: string;
        self: string;
      };
      data: RelationshipData[];
    };
    returns: {
      links: {
        related: string;
        self: string;
      };
      data: RelationshipData[];
    };
  };
  links: {
    self: string;
  };
}

interface RelationshipData {
  type: string;
  id: string;
}

interface Relationship {
  links: {
    related: string;
    self: string;
  };
  data: RelationshipData | null;
}

interface Attributes {
  number: string;
  status_text: string;
  shipping: number;
  tax: number;
  discount: number;
  total: number;
  ready_to_ship: number;
  ready_to_pick: number;
  is_wholesale: null | boolean;
  fraud_hold: number;
  address_hold: number;
  payment_hold: number;
  operator_hold: number;
  allow_partial: number;
  ordered_at: string;
  updated_at: string;
  fulfilled_at: null | string;
  cancelled_at: null | string;
  archived_at: null | string;
  hold_until: null | string;
  ship_before: null | string;
  scheduled_delivery: null | string;
  external_id: string;
  packing_note: string;
  shipping_method_name: string;
  shipping_method_code: string;
  tote: string;
  tags: string;
  created_at: string;
}

interface Included {
  type: string;
  id: string;
  attributes: {
    [key: string]: any;
  };
  relationships?: {
    [key: string]: {
      data: RelationshipData;
    };
  };
}
