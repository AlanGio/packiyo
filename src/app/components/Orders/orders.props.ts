interface OrderAttributes {
  number: string;
  status_text: string;
  shipping: number;
  tax: number;
  discount: number;
  total: number;
  order_channel_name: string;
}

export interface OrderData {
  type: string;
  id: string;
  attributes: OrderAttributes;
}

interface OrdersMetaPage {
  currentPage: number;
  from: number;
  lastPage: number;
  perPage: number;
  to: number;
  total: number;
}

interface OrdersMeta {
  page: OrdersMetaPage;
}

interface OrdersJsonApi {
  version: string;
}

interface OrdersLinks {
  first: string;
  last: string;
}

export interface Orders {
  meta: OrdersMeta;
  jsonapi: OrdersJsonApi;
  links: OrdersLinks;
  data: OrderData[];
}

interface ContactInformation {
  name: string;
  company_name: string;
  address: string;
  address2: string;
  city: string;
  state: string;
  zip: string;
  country: string;
  email: string;
  phone: string;
}

interface OrderItem {
  sku: string;
  quantity: number;
  external_id: string;
  price: number;
}

interface OrderCreateAttributes {
  number: string;
  order_channel_name: string;
  ordered_at: string;
  hold_until: string | null;
  ship_before: string | null;
  external_id: number;
  shipping: number;
  tax: number;
  discount: number;
  packing_note: string;
  shipping_method_name: string;
  shipping_method_code: string;
  tags: string;
  shipping_contact_information_data: ContactInformation;
  billing_contact_information_data: ContactInformation;
  order_item_data: OrderItem[];
}

export interface OrderCreateData {
  type: string;
  attributes: OrderCreateAttributes;
  relationships: {
    customer: {
      data: {
        type: string;
        id: string;
      };
    };
  };
}
