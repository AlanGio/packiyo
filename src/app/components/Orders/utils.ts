import { OrderCreateData } from "./orders.props";

export const initialValues: OrderCreateData = {
  type: "orders",
  attributes: {
    number: "",
    order_channel_name: "Manual Order",
    ordered_at: "2024-05-15 12:00:00",
    hold_until: null,
    ship_before: null,
    external_id: 0,
    shipping: 5.99,
    tax: 50,
    discount: 10,
    packing_note: "Smile while packing!",
    shipping_method_name: "SecondDay",
    shipping_method_code: "0006",
    tags: "wooden, priority",
    shipping_contact_information_data: {
      name: "Sarah Johnson",
      company_name: "",
      address: "123 Maple Street",
      address2: "",
      city: "Anytown",
      state: "NY",
      zip: "12345",
      country: "US",
      email: "api@packiyo.com",
      phone: "555-123-4567",
    },
    billing_contact_information_data: {
      name: "Sarah Johnson",
      company_name: "",
      address: "123 Maple Street",
      address2: "",
      city: "Anytown",
      state: "NY",
      zip: "12345",
      country: "US",
      email: "api@packiyo.com",
      phone: "555-123-4567",
    },
    order_item_data: [
      {
        sku: "40224908181550",
        quantity: 10,
        external_id: "your-internal-order-item-id",
        price: 59.99,
      },
    ],
  },
  relationships: {
    customer: {
      data: {
        type: "customers",
        id: "18",
      },
    },
  },
};

export const generateRndId = () => {
  return Math.round(Math.random() * (999999 - 100000) + 100000);
};
