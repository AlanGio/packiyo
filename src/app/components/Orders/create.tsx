import React, { useState } from "react";
import {
  Container,
  TextField,
  Button,
  Grid,
  Typography,
  Link,
  Alert,
  Snackbar,
} from "@mui/material";
import { OrderCreateData } from "./orders.props";
import { useRouter } from "next/navigation";
import { generateRndId, initialValues } from "./utils";
import { serviceCall } from "@/app/utils/services";

const OrderCreate: React.FC = () => {
  const router = useRouter();
  const [message, setMessage] = useState<{
    type: "error" | "success";
    message: string;
  } | null>(null);
  const [orderData, setOrderData] = useState<OrderCreateData>({
    ...initialValues,
    attributes: {
      ...initialValues.attributes,
      number: generateRndId().toString(),
      external_id: generateRndId(),
    },
  });

  const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const { name, value } = e.target;
    setOrderData((prevData) => ({
      ...prevData,
      attributes: {
        ...prevData.attributes,
        [name]: value,
      },
    }));
  };

  const handleSubmit = (e: React.FormEvent) => {
    e.preventDefault();

    serviceCall(
      `/orders`,
      (response) => {
        if (response instanceof Error) {
          setMessage({
            type: "error",
            message: "Order creation failed!",
          });
        } else {
          setMessage({
            type: "success",
            message: "Order created successfully!",
          });
          router.push(`/order/${response.id}`);
        }
      },
      { data: JSON.stringify({ data: orderData }) },
      "post"
    );
  };

  const handleCloseSnackbar = () => {
    setMessage(null);
  };

  return (
    <Container>
      <Link href="/">Back</Link>
      <Typography variant="h4" gutterBottom>
        Create Order
      </Typography>
      <form onSubmit={handleSubmit}>
        <Grid container spacing={3}>
          <Grid item xs={12}>
            <TextField
              label="Order Number"
              name="number"
              value={orderData.attributes.number}
              onChange={handleChange}
              fullWidth
            />
          </Grid>

          <Grid item xs={12}>
            <TextField
              multiline
              rows={3}
              label="Packing Note"
              name="packing_note"
              value={orderData.attributes.packing_note}
              onChange={handleChange}
              fullWidth
            />
          </Grid>
          <Grid item xs={12}>
            <TextField
              label="Shipping Method Name"
              name="shipping_method_name"
              value={orderData.attributes.shipping_method_name}
              onChange={handleChange}
              fullWidth
            />
          </Grid>
          <Grid item xs={12}>
            <TextField
              label="Shipping Method Code"
              name="shipping_method_code"
              value={orderData.attributes.shipping_method_code}
              onChange={handleChange}
              fullWidth
            />
          </Grid>
          <Grid item xs={12}>
            <TextField
              label="Tags"
              name="tags"
              value={orderData.attributes.tags}
              onChange={handleChange}
              fullWidth
            />
          </Grid>

          {/* Add more fields for shipping and billing contact information as needed */}
          <Grid item xs={12}>
            <Button type="submit" variant="contained" color="primary">
              Create Order
            </Button>
          </Grid>
        </Grid>
      </form>

      <Snackbar
        open={!!message?.message}
        autoHideDuration={6000}
        onClose={handleCloseSnackbar}
        anchorOrigin={{ vertical: "top", horizontal: "right" }}
      >
        <Alert
          onClose={handleCloseSnackbar}
          severity={message?.type}
          sx={{ width: "100%" }}
        >
          {message?.message}
        </Alert>
      </Snackbar>
    </Container>
  );
};

export default OrderCreate;
