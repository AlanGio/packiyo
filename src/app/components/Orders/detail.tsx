import React, { useEffect, useState } from "react";
import {
  Container,
  Typography,
  Grid,
  Paper,
  CircularProgress,
} from "@mui/material";
import { OrderDetailProps } from "./order.detail.props";
import Link from "next/link";
import { serviceCall } from "@/app/utils/services";

const OrderDetail: React.FC<{ orderId: string }> = ({ orderId }) => {
  const [orderData, setOrderData] = useState<OrderDetailProps | null>(null);
  const [loading, setLoading] = useState<boolean>(true);

  useEffect(() => {
    serviceCall(`/orders/${orderId}`, (response) => {
      if (response instanceof Error) {
        setOrderData(null);
      } else {
        setOrderData(response);
      }
      setLoading(false);
    });
  }, [orderId]);

  if (loading) {
    return <CircularProgress />;
  }

  if (!orderData) {
    return <Typography variant="h6">Order not found</Typography>;
  }

  const { attributes, relationships } = orderData;

  return (
    <Container>
      <Link href="/">Back</Link>
      <Typography variant="h4" gutterBottom>
        Order Details
      </Typography>
      <Paper elevation={3} style={{ padding: "16px" }}>
        <Grid container spacing={3}>
          <Grid item xs={12}>
            <Typography variant="h6">
              Order Number: {attributes.number}
            </Typography>
          </Grid>
          <Grid item xs={12}>
            <Typography variant="h6">
              Status: {attributes.status_text}
            </Typography>
          </Grid>
          <Grid item xs={12}>
            <Typography variant="h6">
              Shipping: ${attributes.shipping}
            </Typography>
          </Grid>
          <Grid item xs={12}>
            <Typography variant="h6">Tax: ${attributes.tax}</Typography>
          </Grid>
          <Grid item xs={12}>
            <Typography variant="h6">
              Discount: ${attributes.discount}
            </Typography>
          </Grid>
          <Grid item xs={12}>
            <Typography variant="h6">Total: ${attributes.total}</Typography>
          </Grid>
          <Grid item xs={12}>
            <Typography variant="h6">
              Ordered At: {new Date(attributes.ordered_at).toLocaleString()}
            </Typography>
          </Grid>
          <Grid item xs={12}>
            <Typography variant="h6">
              Updated At: {new Date(attributes.updated_at).toLocaleString()}
            </Typography>
          </Grid>
          <Grid item xs={12}>
            <Typography variant="h6">
              Packing Note: {attributes.packing_note}
            </Typography>
          </Grid>
          <Grid item xs={12}>
            <Typography variant="h6">
              Shipping Method: {attributes.shipping_method_name}
            </Typography>
          </Grid>
          <Grid item xs={12}>
            <Typography variant="h6">Tags: {attributes.tags}</Typography>
          </Grid>
          <Grid item xs={12}>
            <Typography variant="h6">Shipping Information</Typography>
            <Typography>
              Name: {relationships.shipping_contact_information.data?.id}
            </Typography>
          </Grid>
          <Grid item xs={12}>
            <Typography variant="h6">Billing Information</Typography>
            <Typography>
              Name: {relationships.billing_contact_information.data?.id}
            </Typography>
          </Grid>
          <Grid item xs={12}>
            <Typography variant="h6">Order Items</Typography>
            {relationships.order_items.data.map((item) => (
              <Paper
                key={item.id}
                style={{ padding: "8px", marginBottom: "8px" }}
              >
                <Typography>Item ID: {item.id}</Typography>
              </Paper>
            ))}
          </Grid>
        </Grid>
      </Paper>
    </Container>
  );
};

export default OrderDetail;
