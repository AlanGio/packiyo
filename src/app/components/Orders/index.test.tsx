import React from "react";
import { getAllByRole, render } from "@testing-library/react";
import Orders from "./index";

describe("Orders component", () => {
  const orders = [
    {
      id: 1,
      attributes: {
        number: "123",
        status_text: "Pending",
        shipping: 10.0,
        tax: 2.5,
        discount: 5.0,
        total: 17.5,
      },
    },
    {
      id: 2,
      attributes: {
        number: "456",
        status_text: "Completed",
        shipping: 15.0,
        tax: 3.75,
        discount: 7.5,
        total: 26.25,
      },
    },
  ];

  test("renders orders correctly", () => {
    const { getByText } = render(<Orders orders={orders} />);

    orders.forEach((order) => {
      const orderNumber = getByText(order.attributes.number);
      const orderStatus = getByText(`Status: ${order.attributes.status_text}`);
      const orderShipping = getByText(
        `Shipping: $${order.attributes.shipping.toFixed(2)}`
      );
      const orderTax = getByText(`Tax: $${order.attributes.tax.toFixed(2)}`);
      const orderDiscount = getByText(
        `Discount: $${order.attributes.discount.toFixed(2)}`
      );
      const orderTotal = getByText(
        `Total: $${order.attributes.total.toFixed(2)}`
      );

      expect(orderNumber).toMatchSnapshot();
      expect(orderStatus).toMatchSnapshot();
      expect(orderShipping).toMatchSnapshot();
      expect(orderTax).toMatchSnapshot();
      expect(orderDiscount).toMatchSnapshot();
      expect(orderTotal).toMatchSnapshot();
    });
  });
});
