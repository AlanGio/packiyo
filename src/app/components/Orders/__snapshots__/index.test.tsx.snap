// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Orders component renders orders correctly 1`] = `
<h6
  class="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
>
  123
</h6>
`;

exports[`Orders component renders orders correctly 2`] = `
<p
  class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
>
  Status: 
  Pending
</p>
`;

exports[`Orders component renders orders correctly 3`] = `
<p
  class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
>
  Shipping: $
  10.00
</p>
`;

exports[`Orders component renders orders correctly 4`] = `
<p
  class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
>
  Tax: $
  2.50
</p>
`;

exports[`Orders component renders orders correctly 5`] = `
<p
  class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
>
  Discount: $
  5.00
</p>
`;

exports[`Orders component renders orders correctly 6`] = `
<p
  class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
>
  Total: $
  17.50
</p>
`;

exports[`Orders component renders orders correctly 7`] = `
<h6
  class="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
>
  456
</h6>
`;

exports[`Orders component renders orders correctly 8`] = `
<p
  class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
>
  Status: 
  Completed
</p>
`;

exports[`Orders component renders orders correctly 9`] = `
<p
  class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
>
  Shipping: $
  15.00
</p>
`;

exports[`Orders component renders orders correctly 10`] = `
<p
  class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
>
  Tax: $
  3.75
</p>
`;

exports[`Orders component renders orders correctly 11`] = `
<p
  class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
>
  Discount: $
  7.50
</p>
`;

exports[`Orders component renders orders correctly 12`] = `
<p
  class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
>
  Total: $
  26.25
</p>
`;
